Move snapshot APIs from v1beta1 to v1

xing-yang opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR moves Snapshot APIs from v1beta1 to v1.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Move Snapshot APIs to V1.
```

xing-yang commented:
assign @yuxiangqian

xing-yang commented:
Note: Changes to make the validation more strict is not done yet.

yuxiangqian commented:
per offline discussion, this PR will need further refinement.
In order to move to snapshot API to V1, a transition period is needed, more details can be found in this [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1900-volume-snapshot-validation-webhook). In short:
phase 1: introducing validation webhook service, no API change (this has been done)
phase 2: CRD have both v1beta1 and v1 schema, v1beta1 stored in API server, serving both v1 and v1beta1
phase 3: v1 stored (potentially might still need to serve both v1 & v1beta1)

xing-yang commented:
> per offline discussion, this PR will need further refinement.
> In order to move to snapshot API to V1, a transition period is needed, more details can be found in this KEP. In short:
> phase 1: introducing validation webhook service, no API change (this has been done)
> phase 2: CRD have both v1beta1 and v1 schema, v1beta1 stored in API server, serving both v1 and v1beta1
> phase 3: v1 stored (potentially might still need to serve both v1 & v1beta1)

@yuxiangqian I have addressed your comments and made changes for phase 2:  CRDs now have both v1beta1 and v1 schema, v1beta1 stored in API server, serving both v1 and v1beta1.

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+
+	// status represents the current information of a snapshot.
+	// NOTE: status can be modified by sources other than system controllers,
+	// and must not be depended upon for accuracy.
```
What does this mean?

xing-yang replied:
Note: remove this note as it is for developers.

xing-yang replied:
Done

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	// status represents the current information of a snapshot.
+	// NOTE: status can be modified by sources other than system controllers,
+	// and must not be depended upon for accuracy.
+	// Controllers should only use information from the VolumeSnapshotContent object
+	// after verifying that the binding is accurate and complete.
+	// +optional
```
How about:
```suggestion
	// Consumers must verify binding between VolumeSnapshot and
	// VolumeSnapshotContent objects is successful (by validating that both
	// VolumeSnapshot and VolumeSnapshotContent point at each other) before using
	// this object.
	// +optional
```

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	// If not specified, the default snapshot class will be used if one exists.
+	// If not specified, and there is no default snapshot class, dynamic snapshot creation will fail.
+	// Empty string is not allowed for this field.
+	// TODO(xiangqian): a webhook validation on empty string.
```
Is this still a todo? If so, can you add a tracking issue number along with the TODO.

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	Source VolumeSnapshotSource `json:"source" protobuf:"bytes,1,opt,name=source"`
+
+	// volumeSnapshotClassName is the name of the VolumeSnapshotClass requested by the VolumeSnapshot.
+	// If not specified, the default snapshot class will be used if one exists.
+	// If not specified, and there is no default snapshot class, dynamic snapshot creation will fail.
```
Prior to GA, let's improve the defaulting logic for snapshot, such that a given cluster could have multiple default SnapshotClasses: one default per CSI Driver. If a `VolumeSnapshot` does not specify a `SnapshotClass`, look at the `VolumeSnapshotSource` to figure out what the associated CSI Driver is, and use the default `SnapshotClass` associated with that CSI Driver. If more than one `SnapshotClass` exist for a given CSI Driver and more than one have been marked as default, CreateSnapshot should fail and generate an event (similar to CreateVolume failing if multiple StorageClasses are marked as default).

The comments here should be updated to reflect that defaulting behavior.

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+// TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotSource members
+// will not be updated once specified.
+type VolumeSnapshotSource struct {
+	// persistentVolumeClaimName specifies the name of the PersistentVolumeClaim
+	// object in the same namespace as the VolumeSnapshot object where the
+	// snapshot should be dynamically taken from.
```
How about:
```suggestion
	// persistentVolumeClaimName specifies the name of the PersistentVolumeClaim
	// object representing the volume from which a snapshot should be created.
	// This PVC is assumed to be in the same namespace as the VolumeSnapshot
	// object.
	// This field should be set if the snapshot does not exists, and should be
	// created.
```

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	// +optional
+	PersistentVolumeClaimName *string `json:"persistentVolumeClaimName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+	// volumeSnapshotContentName specifies the name of a pre-existing VolumeSnapshotContent
+	// object.
```
How about:
```suggestion
	// volumeSnapshotContentName specifies the name of a pre-existing VolumeSnapshotContent
	// object representing an existing volume snapshot.
	// This field should be set if the snapshot already exists.
```

xing-yang replied:
Done

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+
+// VolumeSnapshotStatus is the status of the VolumeSnapshot
+type VolumeSnapshotStatus struct {
+	// boundVolumeSnapshotContentName represents the name of the VolumeSnapshotContent
+	// object to which the VolumeSnapshot object is bound.
+	// If not specified, it indicates that the VolumeSnapshot object has not been
+	// successfully bound to a VolumeSnapshotContent object yet.
+	// NOTE: Specified boundVolumeSnapshotContentName alone does not mean binding
+	//       is valid. Controllers MUST always verify bidirectional binding between
+	//       VolumeSnapshot and VolumeSnapshotContent to avoid possible security issues.
+	// +optional
```
```suggestion
	// boundVolumeSnapshotContentName is the name of the VolumeSnapshotContent
	// object to which this VolumeSnapshot object intends to bind to.
	// If not specified, it indicates that the VolumeSnapshot object has not been
	// successfully bound to a VolumeSnapshotContent object yet.
	// NOTE: To avoid possible security issues, consumers must verify binding between
	// VolumeSnapshot and VolumeSnapshotContent objects is successful (by validating that
	// both VolumeSnapshot and VolumeSnapshotContent point at each other) before using
	// this object.
	// +optional
```

xing-yang replied:
Done

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+// +kubebuilder:printcolumn:name="SourceSnapshotContent",type=string,JSONPath=`.spec.source.volumeSnapshotContentName`,description="Name of the VolumeSnapshotContent which represents a pre-provisioned snapshot."
+// +kubebuilder:printcolumn:name="RestoreSize",type=string,JSONPath=`.status.restoreSize`,description="Represents the complete size of the snapshot."
+// +kubebuilder:printcolumn:name="SnapshotClass",type=string,JSONPath=`.spec.volumeSnapshotClassName`,description="The name of the VolumeSnapshotClass requested by the VolumeSnapshot."
+// +kubebuilder:printcolumn:name="SnapshotContent",type=string,JSONPath=`.status.boundVolumeSnapshotContentName`,description="The name of the VolumeSnapshotContent to which this VolumeSnapshot is bound."
```
How about `Name of the VolumeSnapshotContent object to which the VolumeSnapshot object intends to bind to. Please note that verification of binding actually requires checking both VolumeSnapshot and VolumeSnapshotContent to ensure both are pointing at each other. Binding MUST be verified prior to usage of this object.`

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="ReadyToUse",type=boolean,JSONPath=`.status.readyToUse`,description="Indicates if a snapshot is ready to be used to restore a volume."
```
```suggestion
// +kubebuilder:printcolumn:name="ReadyToUse",type=boolean,JSONPath=`.status.readyToUse`,description="Indicates if the snapshot is ready to be used to restore a volume."
```

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+// +kubebuilder:printcolumn:name="ReadyToUse",type=boolean,JSONPath=`.status.readyToUse`,description="Indicates if a snapshot is ready to be used to restore a volume."
+// +kubebuilder:printcolumn:name="SourcePVC",type=string,JSONPath=`.spec.source.persistentVolumeClaimName`,description="Name of the source PVC from where a dynamically taken snapshot will be created."
+// +kubebuilder:printcolumn:name="SourceSnapshotContent",type=string,JSONPath=`.spec.source.volumeSnapshotContentName`,description="Name of the VolumeSnapshotContent which represents a pre-provisioned snapshot."
+// +kubebuilder:printcolumn:name="RestoreSize",type=string,JSONPath=`.status.restoreSize`,description="Represents the complete size of the snapshot."
```
```suggestion
// +kubebuilder:printcolumn:name="RestoreSize",type=string,JSONPath=`.status.restoreSize`,description="Represents the minimum size of volume required to create a volume from this snapshot."
```

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+// +kubebuilder:printcolumn:name="RestoreSize",type=string,JSONPath=`.status.restoreSize`,description="Represents the complete size of the snapshot."
+// +kubebuilder:printcolumn:name="SnapshotClass",type=string,JSONPath=`.spec.volumeSnapshotClassName`,description="The name of the VolumeSnapshotClass requested by the VolumeSnapshot."
+// +kubebuilder:printcolumn:name="SnapshotContent",type=string,JSONPath=`.status.boundVolumeSnapshotContentName`,description="The name of the VolumeSnapshotContent to which this VolumeSnapshot is bound."
+// +kubebuilder:printcolumn:name="CreationTime",type=date,JSONPath=`.status.creationTime`,description="Timestamp when the point-in-time snapshot is taken by the underlying storage system."
```
```suggestion
// +kubebuilder:printcolumn:name="CreationTime",type=date,JSONPath=`.status.creationTime`,description="Timestamp when the point-in-time snapshot was taken by the underlying storage system."
```

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	// +optional
+	BoundVolumeSnapshotContentName *string `json:"boundVolumeSnapshotContentName,omitempty" protobuf:"bytes,1,opt,name=boundVolumeSnapshotContentName"`
+
+	// creationTime is the timestamp when the point-in-time snapshot is taken
```
```suggestion
	// creationTime is the timestamp when the point-in-time snapshot was taken
```

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	// +optional
+	CreationTime *metav1.Time `json:"creationTime,omitempty" protobuf:"bytes,2,opt,name=creationTime"`
+
+	// readyToUse indicates if a snapshot is ready to be used to restore a volume.
```
```suggestion
	// readyToUse indicates if the snapshot is ready to be used to restore a volume.
```

xing-yang replied:
Done

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	// +optional
+	ReadyToUse *bool `json:"readyToUse,omitempty" protobuf:"varint,3,opt,name=readyToUse"`
+
+	// restoreSize represents the complete size of the snapshot in bytes.
```
```suggestion
	// restoreSize represents the minimum size of volume required to create a volume from
	// this snapshot.
```

xing-yang replied:
Done

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="ReadyToUse",type=boolean,JSONPath=`.status.readyToUse`,description="Indicates if a snapshot is ready to be used to restore a volume."
```
```suggestion
// +kubebuilder:printcolumn:name="ReadyToUse",type=boolean,JSONPath=`.status.readyToUse`,description="Indicates if the snapshot is ready to be used to restore a volume."
```

xing-yang replied:
Done

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	// On Unix, the command `date +%s%N` returns the current time in nanoseconds
+	// since 1970-01-01 00:00:00 UTC.
+	// +optional
+	CreationTime *int64 `json:"creationTime,omitempty" protobuf:"varint,2,opt,name=creationTime"`
```
We've had this discussion before, but I want to revisit it one last time. Why do we need `CreationTime`, `RestoreSize`, `ReadyToUse`, and `Error` duplicated on both the `VolumeSnapshot` and the `VolumeSnapshotContent` objects?

If we agree the duplication makes sense, please make sure to apply all the comments I left for these objects on the `VolumeSnapshot` side here as well.

xing-yang replied:
TODO: Document why we have them in both objects. 
- status on content can be used for filtering when importing a volumesnapshot.
- volumsnapshot status is used by end users because they can't see volumesnapshotcontent
- Add how they are set and how they are synced.  Eventual consistency.

xing-yang replied:
Done

xing-yang replied:
Done

saad-ali commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+
+	// name of the VolumeSnapshotClass to which this snapshot belongs.
+	// +optional
+	VolumeSnapshotClassName *string `json:"volumeSnapshotClassName,omitempty" protobuf:"bytes,4,opt,name=volumeSnapshotClassName"`
```
Should we create a Parameters field and copy the `SnapshotClass.parameters` field here. A snapshot must be able to operate independently of the `SnapshotClass` after provisioning. We have `DeletionPolicy` here already, which is good, but do we need parameters too?

saad-ali replied:
@msau42 reminded me that PV/PVC actually handles this in a better way. Instead of copying all parameters over from the Class object (which may include lots of unnecessary information), the CSI `CreateVolumeResponse` contains a set of `volume_context` -- basically letting the driver decide what additional important info should be maintained about the volume. We should do something similar here. See https://github.com/container-storage-interface/spec/blob/master/csi.proto#L488

xing-yang replied:
We'll take a look.

msau42 commented on client/hack/update-generated-code.sh:
```diff
@@ -26,7 +26,7 @@ SCRIPT_ROOT=$(unset CDPATH && cd $(dirname "${BASH_SOURCE[0]}")/.. && pwd)
 ${GOPATH}/src/k8s.io/code-generator/generate-groups.sh "deepcopy,client,informer,lister" \
   github.com/kubernetes-csi/external-snapshotter/client/v3 github.com/kubernetes-csi/external-snapshotter/client/v3/apis \
-  volumesnapshot:v1beta1 \
+  volumesnapshot:v1 \
```
Do we want to keep generated client code for both v1beta and v1?

xing-yang replied:
Yes.  I expect the existing generated client code for v1beta to stay, but we will only be changing generated client code for v1.  Therefore this code generating script should only handle v1.  I'll add a note there.

msau42 commented on pkg/validation-webhook/snapshot.go:
```diff
@@ -20,7 +20,7 @@ import (
 	"reflect"
 
-	volumesnapshotv1beta1 "github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1beta1"
+	volumesnapshotv1 "github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1"
```
The webhook needs to support both v1beta and v1

msau42 replied:
Based on discussion today, we're keeping the storage version at v1beta, so the webhook needs to validate v1beta objects

xing-yang replied:
I'll test this.

xing-yang replied:
I've tested this.  You are right, this should still be using volumesnapshotv1beta1.  I modified the code now.  It works as expected.

msau42 commented on client/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml:
```diff
@@ -56,116 +154,63 @@ spec:
+                    description: volumeSnapshotContentName specifies the name of a pre-existing VolumeSnapshotContent object. This field is immutable.
                     type: string
                 type: object
+                oneOf:
```
We shouldn't tighten validation on existing apis. The webhook is supposed to fix this

xing-yang replied:
Sure.  I'll remove this from v1beta1.

xing-yang replied:
Done

msau42 commented on client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml:
```diff
@@ -52,88 +183,50 @@ spec:
                     type: string
                 type: object
+                oneOf:
+                - required: ["snapshotHandle"]
```
Same thing, shouldn't tighten validation on existing apis

xing-yang replied:
Sure.  I'll remove this from v1beta1.

xing-yang replied:
Done

xing-yang commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="ReadyToUse",type=boolean,JSONPath=`.status.readyToUse`,description="Indicates if a snapshot is ready to be used to restore a volume."
+// +kubebuilder:printcolumn:name="SourcePVC",type=string,JSONPath=`.spec.source.persistentVolumeClaimName`,description="Name of the source PVC from where a dynamically taken snapshot will be created."
```
Done

xing-yang commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+
+	// readyToUse indicates if a snapshot is ready to be used to restore a volume.
+	// In dynamic snapshot creation case, this field will be filled in with the
+	// "ready_to_use" value returned from CSI "CreateSnapshotRequest" gRPC call.
```
Done

xing-yang commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	CreationTime *metav1.Time `json:"creationTime,omitempty" protobuf:"bytes,2,opt,name=creationTime"`
+
+	// readyToUse indicates if a snapshot is ready to be used to restore a volume.
+	// In dynamic snapshot creation case, this field will be filled in with the
```
Done

xing-yang commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:
```diff
@@ -0,0 +1,395 @@
+	// +optional
+	RestoreSize *resource.Quantity `json:"restoreSize,omitempty" protobuf:"bytes,4,opt,name=restoreSize"`
+
+	// error is the last observed error during snapshot creation, if any.
```
Done.

yuxiangqian commented on client/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml:
```diff
@@ -18,35 +18,133 @@ spec:
       type: boolean
-    - description: Name of the source PVC from where a dynamically taken snapshot
-        will be created.
+    - description: If a new snapshot needs to be created, this contains the name of the source PVC from which this snapshot was (or will be) created.
```
I actually prefer the original description better, its shorter and clear enough ;)

xing-yang replied:
This is changed to address Saad's comments.

yuxiangqian commented on client/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml:
```diff
@@ -18,35 +18,133 @@ spec:
       type: string
-    - description: Name of the VolumeSnapshotContent which represents a pre-provisioned
-        snapshot.
+    - description: If a snapshot already exists, this contains the name of the existing VolumeSnapshotContent object representing the existing snapshot.
```
ditto

xing-yang replied:
This is changed to address Saad's comments.

yuxiangqian commented on client/config/crd/snapshot.storage.k8s.io_volumesnapshots.yaml:
```diff
@@ -18,35 +18,133 @@ spec:
       name: SourceSnapshotContent
       type: string
-    - description: Represents the complete size of the snapshot.
+    - description: Represents the minimum size of volume required to create a volume from this snapshot.
```
how about? "to rehydrate from this snapshot"

xing-yang replied:
Sure

yuxiangqian commented on vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/clientset/versioned/typed/volumesnapshot/v1/volumesnapshotcontent.go:
```diff
@@ -0,0 +1,184 @@
+/*
```
I am having a hard time to understand these generated files, taking:
clientset/versioned/typed/volumesnapshot/v1/volumesnapshotcontent.go as an example, there seems to be two identical ones generated:
vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/clientset/versioned/typed/volumesnapshot/v1/volumesnapshotcontent.go
and 
client/clientset/versioned/typed/volumesnapshot/v1/volumesnapshotcontent.go

could you shed some light what are them for?

xing-yang replied:
Actually only `client/clientset/versioned/typed/volumesnapshot/v1/volumesnapshotcontent.go` is generated by code-generator. `vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/clientset/versioned/typed/volumesnapshot/v1/volumesnapshotcontent.go` is generated by running "go mod vendor" because we have a external-snapshotter/vendor folder that already has those files for v1beta1.

yuxiangqian commented on client/hack/README.md:
```diff
@@ -8,7 +8,7 @@ This README documents:
 This is the script to update clientset/informers/listers and API deepcopy code using [code-generator](https://github.com/kubernetes/code-generator).
 
-Make sure to run this script after making changes to /client/apis/volumesnapshot/v1beta1/types.go.
+Make sure to run this script after making changes to /client/apis/volumesnapshot/v1/types.go.
```
My understanding is that v1beta1/types.go still exists. This PR only add a v1/types.go into the repo is that right?

xing-yang replied:
Yes, but we no longer makes changes to v1beta1.

yuxiangqian commented on client/hack/README.md:
````diff
@@ -95,3 +99,68 @@ Update the restoreSize property to use type string only:
 
 ```
 
+* Because we need to serve both v1 and v1beta1 snapshot APIs, we need to make a copy of the generated v1 schema, paste it below and change the version to v1beta1 under `versions:` of each generated yaml file. See `snapshot.storage.k8s.io_volumesnapshots.yaml` as an example. `served` is true for both v1beta1 and v1. `storage` is true for v1beta and false for v1.
````
I am not sure I fully understand this phrase:
"Because we need to serve both v1 and v1beta1 snapshot APIs, we need to make a copy of the generated v1 schema, paste it below and change the version to v1beta1 under `versions:` of each generated yaml file."

Do you mean after the v1 yaml files have been generated., the 'spec' piece needs to be copied to its corresponding yaml file? what does "change to v1beta1 under 'versions:'" mean? change v1. to v1beta1?

xing-yang replied:
I mean we need to copy the generated v1 section and pasted below for v1beta1.  Maybe we should just copy from the originally generated v1beta1 manifest?

yuxiangqian reviewed (commented):
some nits, overall looks good! thanks Xing

xing-yang commented on client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml:
```diff
@@ -52,88 +183,47 @@ spec:
               volumeSnapshotClassName:
-                description: name of the VolumeSnapshotClass to which this snapshot
-                  belongs.
+                description: name of the VolumeSnapshotClass to which this snapshot belongs.
```
this is automatically generated.

xing-yang commented:
@saad-ali @msau42 @yuxiangqian addressed your comments.  PTAL.

xing-yang commented:
Note: unit test failed because I manually modified release-tools to install the V1 snapshot CRDs.

xing-yang commented:
/hold

xing-yang commented:
This is addressed by different PRs.  Closing this one.
